### PR TITLE
API to generate report in PDF/PNG format

### PR DESCRIPTION
**Pull request by zhongnansu**
*Issue #, if available:*
#11 
*Description of changes:*
- API endpoint `http://localhost:5601/api/reporting/generateReport`
- API request body as json:
```
{
    "url": "http://localhost:5601/app/kibana#/dashboard/7adfa750-4c81-11e8-b3d7-01146121b73d?"
    "itemName": "Flights_Dashboard",
    "reportFormat": "png",
    "windowWidth": 2000,
    "windowLength": 500
}

```
- API response: a file saved locally
```
attachment; filename=Flights_Dashboard_2020-06-17T17%3A29%3A41.152Z_205f0910-b0c0-11ea-b51b-7f924012e825.png
```

- Refactored the server side codebase
- Added unit tests for utility functions
- Added .gitignore
- Added prettier config file

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: server/models/types.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+export type ServerResponse<T> = { ok: false; error: string } | { ok: true; response: T };
```
Comment (davidcui-amzn) [inline server/models/types.ts:16]: np: new line

File: server/routes/dashboard.js [hunk 2]
```diff
@@ -13,38 +13,33 @@
...
-  function getJson() {
-      var url = "http://localhost:9200/.kibana/_search?q=type:dashboard";
-      var HttpRequest = new XMLHttpRequest();
-      HttpRequest.open("GET", url, false);
-      HttpRequest.send(null);
-      return HttpRequest.responseText; 
-  }
\ No newline at end of file
+function getJson() {
+  var url = 'http://localhost:9200/.kibana/_search?q=type:dashboard';
```
Comment (davidcui-amzn) [inline server/routes/dashboard.js:40]: I think we shouldn't hardcode `localhost:9200` for Elasticsearch- use `server.plugins.elasticsearch.getCluster('data');` instead

File: server/routes/generateReport.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+import { Server } from 'hapi';
+import * as Joi from '@hapi/joi';
```
Comment (davidcui-amzn) [inline server/routes/generateReport.ts:17]: Can we avoid the wild card import here?

File: server/routes/scheduler.js [hunk 4]
```diff
@@ -13,28 +13,19 @@
...
+    path: '/api/reporting/schedule',
+    method: 'POST',
+    async handler(response) {
+      try {
+        await schedule_job();
+      } catch (e) {
+        console.log(e);
       }
-    });
+      return "hello client, it's your friend server";
```
Comment (davidcui-amzn) [inline server/routes/scheduler.js:26]: I wrote this response when I was just testing on my local machine, I think we should try to return a more meaningful API response.

File: server/services/generateReportService.ts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+import { Legacy } from 'kibana';
+import { Server, Request, ResponseToolkit } from 'hapi';
+import { RequestParams } from '@elastic/elasticsearch';
+import puppeteer from 'puppeteer';
+import imagesToPdf from 'images-to-pdf';
+import fs from 'fs';
+import { v1 as uuidv1 } from 'uuid';
+
+// TODO: change the following accordingly
+// import { AcknowledgedResponse, AddPolicyResponse, AddResponse, CatIndex, GetIndicesResponse, SearchResponse } from "../models/interfaces";
```
Comment (davidcui-amzn) [inline server/services/generateReportService.ts:26]: Can we remove these comments in this file?

File: server/test/generateReport.test.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import 'regenerator-runtime/runtime';
```
Comment (davidcui-amzn) [inline server/test/generateReport.test.ts:1]: Copyright notice is missing

File: server/test/generateReport.test.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      input.windowWidth,
+      input.windowLength
+    );
+    expect(fileName).toContain(input.itemName + '_' + timeCreated);
+  }, 20000);
+
+  test('generate PDF success', async () => {
+    expect.assertions(1);
+    const input = {
+      url: 'https://www.google.com/',
```
Comment (seraphjiang) [inline server/test/generateReport.test.ts:36]: could we use other domain? at least domain we own?

File: server/test/generateReport.test.ts (hunk 6 above)
Comment (zhongnansu) [reply server/test/generateReport.test.ts:1]: good catch

File: server/services/generateReportService.ts (hunk 5 above)
Comment (zhongnansu) [reply server/services/generateReportService.ts:26]: will do

File: server/models/types.ts (hunk 1 above)
Comment (zhongnansu) [reply server/models/types.ts:16]: will do

File: server/routes/dashboard.js (hunk 2 above)
Comment (zhongnansu) [reply server/routes/dashboard.js:40]: That's part is not implemented yet. I just leave the code as it is. I think I'll just delete those files for now, will plan those parts for future PRs

File: server/routes/scheduler.js (hunk 4 above)
Comment (zhongnansu) [reply server/routes/scheduler.js:26]: will delete

File: server/routes/dashboard.js (hunk 2 above)
Comment (davidcui-amzn) [reply server/routes/dashboard.js:40]: Sounds good 👍

File: server/routes/dashboard.js [hunk 8]
```diff
@@ -13,38 +13,33 @@
  * permissions and limitations under the License.
  */
 
-import { Server, Request, ResponseToolkit } from 'hapi';
+var XMLHttpRequest = require('xmlhttprequest').XMLHttpRequest;
```
[the remark this answers is not in the document]
Comment (zhongnansu) [reply server/routes/dashboard.js:16]: Good point. `scheduler.js` and `dashboard.js` are from Poc stage. It's deprecated and I will delete them now and plan in future PRs.

File: server/services/generateReportService.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (zhongnansu) [reply server/services/generateReportService.ts:26]: will do

File: server/routes/generateReport.ts (hunk 3 above)
Comment (zhongnansu) [reply server/routes/generateReport.ts:17]: good point

File: server/routes/dashboard.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (zhongnansu) [reply server/routes/dashboard.js:40]: Good point. `scheduler.js` and `dashboard.js` are from Poc stage. It's deprecated and I will delete them now and plan in future PRs.

File: server/services/generateReportService.ts [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+//@ts-ignore
+import { Legacy } from 'kibana';
```
Comment (davidcui-amzn) [inline server/services/generateReportService.ts:17]: If we are going to move away from the Legacy kibana framework, will this import pose a problem after we move away from that?

File: server/test/generateReport.test.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (davidcui-amzn) [inline server/test/generateReport.test.ts:2]: np: Copyright 2020, not 2019

File: server/services/generateReportService.ts (hunk 9 above)
Comment (zhongnansu) [reply server/services/generateReportService.ts:17]: This imports the Elasticsearch Plugin type from Legacy for communicating with ES plugin. Cannot find a solution to not use legacy for now, also referred to other OD plugins. Any suggestions?

File: server/test/generateReport.test.ts (hunk 10 above)
Comment (zhongnansu) [reply server/test/generateReport.test.ts:2]: will fix it

File: server/test/generateReport.test.ts (hunk 7 above)
Comment (zhongnansu) [reply server/test/generateReport.test.ts:51]: will change to https://opendistro.github.io/for-elasticsearch-docs/

File: server/services/generateReportService.ts [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+          url,
+          itemName,
+          windowWidth,
+          windowLength
+        );
+        // TODO: save metadata into ES
+        // const { callWithRequest } = this.esDriver.getCluster(CLUSTER.DATA)
+        // const params: RequestParams.Index = { index: "report", body: { url, itemName, reportFormat, timeCreated } }
+        // await callWithRequest(req, "index", params)
+
```
[the remark this answers is not in the document]
Comment (zhongnansu) [reply server/services/generateReportService.ts:63]: will do

File: server/services/generateReportService.ts (hunk 9 above)
Comment (davidcui-amzn) [reply server/services/generateReportService.ts:17]: No worries if also referred to in other OD plugins- was just worried about potentially accruing additional tech debt.

File: server/services/generateReportService.ts [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        reportFormat,
+        windowWidth = 1200,
+        windowLength = 800,
+      } = req.payload as {
+        url: string;
+        itemName: string;
+        reportFormat: string;
+        windowWidth: number;
+        windowLength: number;
+      };
```
Comment (mihirsoni) [inline server/services/generateReportService.ts:50]: np: Extract this as separate type.

File: server/services/generateReportService.ts [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+
+      if (reportFormat === 'png') {
```
Comment (mihirsoni) [inline server/services/generateReportService.ts:52]: Move `reportFormat` to enum.

File: server/services/generateReportService.ts [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+          windowWidth,
+          windowLength
+        );
+        //@ts-ignore
+        return h.file(fileName + '.pdf', { mode: 'attachment' });
+      }
+
+      return { message: 'no support for such format: ' + reportFormat };
+    } catch (err) {
+      console.error('Reporting-Generate-PDF', err);
```
Comment (mihirsoni) [inline server/services/generateReportService.ts:75]: Use logger instead of `console.log`

File: server/services/generateReportService.ts [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+): Promise<{ timeCreated: string; fileName: string }> => {
+  const { timeCreated, fileName } = await generatePNG(
+    url,
+    itemName,
+    windowWidth,
+    windowLength
+  );
+  //add png to pdf to avoid long page split
+  await imagesToPdf([fileName + '.png'], fileName + '.pdf');
+  return { timeCreated, fileName };
```
Comment (mihirsoni) [inline server/services/generateReportService.ts:101]: Move this under `try` catch

File: server/test/generateReport.test.ts [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    const regex = /[.](png|pdf)$/;
+    fs.readdirSync(path)
+      .filter((f) => regex.test(f))
+      .map((f) => fs.unlinkSync(path + f));
+  });
+
+  test('generate PNG success', async () => {
+    expect.assertions(1);
+    const input = {
+      url: 'https://www.google.com/',
```
Comment (seraphjiang) [inline server/test/generateReport.test.ts:33]: use our own domain

File: server/services/generateReportService.ts (hunk 14 above)
Comment (zhongnansu) [reply server/services/generateReportService.ts:75]: Any logging library that you suggest?

File: server/services/generateReportService.ts [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+  await page.setViewport({
+    width: windowWidth,
+    height: windowLength,
+  });
+
+  // TODO: this element is for Dashboard page, need to think about addition params to select html element with source(Visualization, Dashboard)
+  // const ele = await page.$('div[class="react-grid-layout dshLayout--viewing"]')
+
+  const timeCreated = new Date().toISOString();
+  const fileName = itemName + '_' + timeCreated + '_' + uuidv1();
```
Comment (seraphjiang) [inline server/services/generateReportService.ts:125]: make a function for readability
```
function getFileName(itemName){
  return `${itemName}_${uuidvi()}`
}

"2020-06-18T05:32:23.547Z" ISOString contains ```:``` which is restrict for file name, could we remove it from file name if not necessary
```

File: server/services/generateReportService.ts [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+      } else if (reportFormat === 'pdf') {
+        const { fileName } = await generatePDF(
+          url,
+          itemName,
+          windowWidth,
+          windowLength
+        );
+        //@ts-ignore
+        return h.file(fileName + '.pdf', { mode: 'attachment' });
+      }
```
Comment (seraphjiang) [inline server/services/generateReportService.ts:71]: refactor code - DRY
```
        const { fileName } = await generatePNG(
          url,
          itemName,
          windowWidth,
          windowLength
        );

        return h.file(`${fileName}.${reportFormat}`, { mode: 'attachment' });
```

File: server/services/generateReportService.ts (hunk 14 above)
Comment (zhongnansu) [reply server/services/generateReportService.ts:75]: I also looked into AD implementation, do you think it's good to do the same thing as AD. https://github.com/opendistro-for-elasticsearch/anomaly-detection-kibana-plugin/blob/3ce118c7c2c49bcb859f84beaf763a36bf2ef00e/index.js#L60

File: server/services/generateReportService.ts (hunk 18 above)
Comment (zhongnansu) [reply server/services/generateReportService.ts:71]: Good catch!

File: server/services/generateReportService.ts (hunk 17 above)
Comment (zhongnansu) [reply server/services/generateReportService.ts:125]: ACK

File: server/services/generateReportService.ts (hunk 13 above)
Comment (zhongnansu) [reply server/services/generateReportService.ts:52]: yeah that makes sense

File: server/services/generateReportService.ts (hunk 15 above)
Comment (zhongnansu) [reply server/services/generateReportService.ts:101]: fixed

File: server/test/generateReport.test.ts (hunk 16 above)
Comment (zhongnansu) [reply server/test/generateReport.test.ts:33]: forgot this one, will fix

File: server/services/generateReportService.ts (hunk 14 above)
Comment (mihirsoni) [reply server/services/generateReportService.ts]: There is already logging with Kibana you just need to pass it over and use it.

**Pull request merged**